crates/types/src/consensus.rs

```diff
@@ -835,6 +843,103 @@ impl<TYPES: NodeType> Consensus<TYPES> {
         }
         Some(())
     }
+
+    /// Returns true if the current high qc is for the last block in the epoch
+    pub fn is_high_qc_for_last_block(&self) -> bool {
+        let high_qc = self.high_qc();
+        self.is_qc_for_last_block(high_qc)
+    }
+
+    /// Returns true if the given qc is for the last block in the epoch
+    pub fn is_qc_for_last_block(&self, cert: &QuorumCertificate<TYPES>) -> bool {
+        let Some(leaf) = self.saved_leaves.get(&cert.data().leaf_commit) else {
+            return false;
+        };
+        let block_height = leaf.height();
+        if block_height == 0 || self.epoch_height == 0 {
+            false
+        } else {
+            block_height % self.epoch_height == 0
```

Comment:
There's a potential division by zero issue here. We should probably add a non-zero check before this line.